Fixes #36219 - use YAML.safe_load instead of YAML.load

Ron-Lavi opened the pull request:
as explained in this rubocop rule https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Security/YAMLLoad


<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

ekohl commented on lib/tasks/puppet.rake:
```diff
@@ -46,7 +46,7 @@ namespace :puppet do
-          nodeinfo = YAML.load `#{script} #{host.name}`
+          nodeinfo = YAML.safe_load `#{script} #{host.name}`
           if nodeinfo.is_a?(Hash)
             $stdout.puts "DONE" if host.importNode nodeinfo
```
Shall we drop the task instead? `importNode` was removed in 9584548fbc6188fefd462b6593384aed0d8c20f1. If we want to keep this task (which I seriously doubt), then it should live in foreman_puppet instead.

Ron-Lavi replied:
For easier cherry-picking, I think we should do this in a separate PR, opened Redmine for it: https://projects.theforeman.org/issues/36222
cc @nadjaheitmann

ekohl replied:
You are correct that it deserves its own commit and own issue, but if I'm honest, I feel safer to cherry pick the removal than using `YAML.safe_load`. We know this task is broken since Foreman 3.0 so it's broken anyway. And if you have the permission to call a rake task, you're already in full control of Foreman that you could also call `foreman-rake console`. Using `YAML.safe_load` can break things and https://github.com/rails/rails/commit/611990f1a6c137c2d56b1ba06b27e5d2434dcd6a certainly did.

Note I'm not saying we shouldn't pick, because security can be more important but we should be careful about it.

Ron-Lavi replied:
OK then, dropping it in this PR as a separate commit,
I don't mind cherry-picking it as well

ekohl reviewed (changes requested):
It should be noted that in Ruby 3 the default load is `safe_load`, so starting to use `safe_load` is good for upgrading too.

The test failures on `Symbol` show that we do rely on some permitted classes. I'm just not sure which classes should be permitted.

ronlavi2412 committed:
Fixes #36219 - use YAML.safe_load instead of YAML.load

as explained in this rubocop rule https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Security/YAMLLoad

Ron-Lavi commented:
@ekohl can you have another look?

ekohl commented on app/models/report.rb:
```diff
@@ -62,7 +62,7 @@ def self.inherited(child)
   def metrics
     return {} if self[:metrics].nil?
-    YAML.load(read_metrics).with_indifferent_access
+    YAML.safe_load(read_metrics).with_indifferent_access
```
This just made me realize: `HashWithIndifferentAccess` always stores its keys as strings, so you'll never have a `Symbol`.

ekohl commented on lib/tasks/puppet.rake:
```diff
@@ -1,110 +0,0 @@
-    end
-  end
-
-  namespace :migrate do
-    desc "
-    Migrate Puppet configuration from Host and Hostgroup attributes to parameters.
-    This task is relevant if you want to continue using Puppet provisioning,
-    but do not want to install the Puppet plugin and get the full ENC funcionality."
```
The description clearly states it's intended to be used without the Puppet plugin, so I don't think this should be removed.

Ron-Lavi replied:
dropped just the specific task now, thanks @ekohl !

ekohl commented on lib/tasks/parameters.rake:
```diff
@@ -24,7 +24,7 @@ namespace :parameters do
   task :cast_key_types_and_values => :environment do
     def override_key_type_and_value(param)
       key_type_name = 'string'
```
Unrelated, but line looks redundant.
```suggestion
```

ronlavi2412 committed:
Fixes #36222 - drop unused puppet rake task

ronlavi2412 committed:
Refs #36219 - remove redundant line

in parameters rake task, key_type_name was assigned twice

Ron-Lavi commented:
Ready for another round @ekohl

Ron-Lavi commented:
[test unit]

ekohl reviewed (approved):
I worry that I can't oversee all the issues, but I think this may also be one of the cases where we should merge it and rely on testing. Longer term this is a must since in Psych 4 (Ruby 3.1) `YAML.load` has changed into `YAML.safe_load`.

I also wonder if longer term we should change the Settings value column into JSON columns. That way they're natively stored in PostgreSQL and custom code to (de)serialize.

ekohl commented:
@Ron-Lavi could you rebase and the Refs commit into the original Fixes. I think [Refs #36219 - use permitted_classes](https://github.com/theforeman/foreman/pull/9660/commits/36dfd08dd50f258b998d9fca5ead216a3acf3a1e) is part of the original commit. The other can remain if you want to.

Ron-Lavi commented:
Thanks @ekohl, rebased

ekohl commented:
I can't say if the Katello failures are related or not. Probably not, but I'm a bit cautious with it. Do you happen to know?

ekohl commented:
https://github.com/theforeman/foreman/pull/9777 should be a safe reference point. Let's see what that has as a result.

ekohl commented:
Same failures, so I'm going to merge this now. Thanks for addressing this!

ekohl commented:
I think this broke Puppet reports because we see `Tried to load unspecified class: ActiveSupport::HashWithIndifferentAccess` in the logs:
https://community.theforeman.org/t/foreman-nightly-rpm-pipeline-1955-failed/34285

But I'll admit I didn't look too carefully. It can also be Hammer output, but IIRC Hammer doesn't use ActiveSupport.

ofedoren commented:
> Hammer doesn't use ActiveSupport

Yeap.

And this reminds me https://github.com/theforeman/foreman/pull/9304/files#diff-c1fd91cb1911a0512578b99f657554526f3e1421decdb9e908712beab57e10f9

ehelms commented:
@Ron-Lavi could you have a look at https://github.com/theforeman/foreman/pull/9660#issuecomment-1642576310 ? We may need a follow up to complete this fully.

evgeni commented:
I've opened https://github.com/theforeman/foreman/pull/9782 to fix this

ehelms commented:
Thanks @evgeni for jumpin in.
